data/ai/lua/ca_recruit_rushers.lua

```diff
@@ -1,50 +1,1096 @@
-                ca_castle_switch = wesnoth.require("ai/lua/ca_castle_switch.lua")
-                break
+local AH = wesnoth.require "ai/lua/ai_helper.lua"
+local M = wesnoth.map
+local LS = wesnoth.require "location_set"
+
+math.randomseed(os.time())
+local recruit_data = { turn = wesnoth.current.turn }
+
+
+local function ca_castle_switch()
+    for ai_tag in wml.child_range(wesnoth.sides[wesnoth.current.side].__cfg, 'ai') do
+        for stage in wml.child_range(ai_tag, 'stage') do
+            for ca in wml.child_range(stage, 'candidate_action') do
+                if ca.location and string.find(ca.location, 'ca_castle_switch') then
+                    return wesnoth.dofile("ai/lua/ca_castle_switch.lua")
+                end
+            end
+        end
+    end
+
+    return nil
+end
+
+
+local function no_village_cost(recruit_id)
+    return wesnoth.unit_types[recruit_id].cost+wesnoth.unit_types[recruit_id].level+wesnoth.sides[wesnoth.current.side].village_gold
+end
+
+
+local function get_hp_efficiency(table, recruit_id)
+    -- raw durability is a function of hp and the regenerates ability
+    -- efficiency decreases faster than cost increases to avoid recruiting many expensive units
+    -- there is a requirement for bodies in order to block movement
+
+    -- There is currently an assumption that opponents will average about 15 damage per strike
+    -- and that two units will attack per turn until the unit dies to estimate the number of hp
+    -- gained from regeneration
+    local effective_hp = wesnoth.unit_types[recruit_id].max_hitpoints
+
+    local unit = wesnoth.units.create {
+        type = recruit_id,
+        random_traits = false,
+        name = "X",
+        random_gender = false
+    }
+    -- Find the best regeneration ability and use it to estimate hp regained by regeneration
+    local abilities = wml.get_child(unit.__cfg, "abilities")
+    local regen_amount = 0
+    if abilities then
+        for regen in wml.child_range(abilities, "regenerate") do
+            if regen.value > regen_amount then
+                regen_amount = regen.value
+            end
+        end
+        effective_hp = effective_hp + (regen_amount * effective_hp/30)
+    end
+    local hp_score = math.max(math.log(effective_hp/20),0.01)
+    local efficiency = hp_score/(wesnoth.unit_types[recruit_id].cost^2)
+    local no_village_efficiency = hp_score/(no_village_cost(recruit_id)^2)
+
+    table[recruit_id] = {efficiency, no_village_efficiency}
+    return {efficiency, no_village_efficiency}
+end
+local efficiency = {}
+setmetatable(efficiency, { __index = get_hp_efficiency })
+
+
+local function poisonable(unit)
+    return not unit.status.unpoisonable
+end
+
+
+local function drainable(unit)
+    return not unit.status.undrainable
+end
+
+
+local function get_best_defense(unit)
+    local terrain_archetypes = { "Wo", "Ww", "Wwr", "Ss", "Gt", "Ds", "Ft", "Hh", "Mm", "Vi", "Ch", "Uu", "At", "Qt", "^Uf", "Xt", "Tb" }
```

Comment:
Might as well update the list of terrain codes while you're copying everything.
^Uf and Tb are the same terrain but ^Uf is deprecated, so it can go. I suggest using Tt, which is what both are aliases of.
Mine rails got a special code and while it's not useful in general, there might be some value in adding it (I left it out since it's more of a functional change in the suggestion, but worth considering).
```suggestion
    local terrain_archetypes = { "Wo", "Ww", "Wwr", "Ss", "Gt", "Ds", "Ft", "Hh", "Mm", "Vi", "Ch", "Uu", "At", "Qt", "Xt", "Tt" }
```